src/compiler/crystal/syntax/lexer.cr

```diff
@@ -1456,567 +1450,192 @@ module Crystal
+      # Check or determine suffix
+      if suffix_size == 0
+        raise_value_doesnt_fit_in(negative ? Int64 : UInt64, error_number_string, start) unless @token.value
+        @token.number_kind = case number_size
+                             when 0..9   then :i32
+                             when 10     then raw_number_string.to_i32? ? :i32 : :i64
+                             when 11..18 then :i64
+                             when 19
+                               if raw_number_string.to_i64?
+                                 :i64
+                               elsif negative
+                                 raise_value_doesnt_fit_in(Int64, error_number_string, start)
+                               else
+                                 :u64
+                               end
+                             when 20
+                               raise_value_doesnt_fit_in(Int64, error_number_string, start) if negative
+                               raise_value_doesnt_fit_in(UInt64, error_number_string, start) unless raw_number_string.to_u64?
+                               :u64
+                             else
+                               raise_value_doesnt_fit_in(Int64, error_number_string, start) if negative
+                               raise_value_doesnt_fit_in(UInt64, error_number_string, start)
+                             end
+      else
+        case @token.number_kind
+        when :i8   then gen_check_int_fits_in_size(Int8, :i8, 3, number_size, raw_number_string, error_number_string, current_pos, start, negative)
+        when :u8   then gen_check_int_fits_in_size(UInt8, :u8, 3, number_size, raw_number_string, error_number_string, current_pos, start, negative)
+        when :i16  then gen_check_int_fits_in_size(Int16, :i16, 5, number_size, raw_number_string, error_number_string, current_pos, start, negative)
+        when :u16  then gen_check_int_fits_in_size(UInt16, :u16, 5, number_size, raw_number_string, error_number_string, current_pos, start, negative)
+        when :i32  then gen_check_int_fits_in_size(Int32, :i32, 10, number_size, raw_number_string, error_number_string, current_pos, start, negative)
+        when :u32  then gen_check_int_fits_in_size(UInt32, :u32, 10, number_size, raw_number_string, error_number_string, current_pos, start, negative)
+        when :i64  then gen_check_int_fits_in_size(Int64, :i64, 19, number_size, raw_number_string, error_number_string, current_pos, start, negative)
+        when :u64  then gen_check_int_fits_in_size(UInt64, :u64, 20, number_size, raw_number_string, error_number_string, current_pos, start, negative)
+        when :i128 then gen_check_int_fits_in_size(Int128, :i64, 19, number_size, raw_number_string, error_number_string, current_pos, start, negative, Int64)
+        when :u128 then gen_check_int_fits_in_size(UInt128, :u64, 20, number_size, raw_number_string, error_number_string, current_pos, start, negative, UInt64)
+        end
```

Comment:
It could be refactored to cut down the repetition a bit:

```suggestion
        args = {number_size, raw_number_string, error_number_string, current_pos, start, negative}

        case @token.number_kind
        when :i8   then gen_check_int_fits_in_size(Int8, :i8, 3, *args)
        when :u8   then gen_check_int_fits_in_size(UInt8, :u8, 3, *args)
        when :i16  then gen_check_int_fits_in_size(Int16, :i16, 5, *args)
        when :u16  then gen_check_int_fits_in_size(UInt16, :u16, 5, *args)
        when :i32  then gen_check_int_fits_in_size(Int32, :i32, 10, *args)
        when :u32  then gen_check_int_fits_in_size(UInt32, :u32, 10, *args)
        when :i64  then gen_check_int_fits_in_size(Int64, :i64, 19, *args)
        when :u64  then gen_check_int_fits_in_size(UInt64, :u64, 20, *args)
        when :i128 then gen_check_int_fits_in_size(Int128, :i64, 19, *args, Int64)
        when :u128 then gen_check_int_fits_in_size(UInt128, :u64, 20, *args, UInt64)
        end
```

Reply:
Seems that this diff caused CI to fail (compilation error). So this would need a fixup.
Or, actually, this seems impossible to accept, because `args` is a runtime value whereas `gen_check_int_fits_in_size` is a macro.

Reply:
Good point, unfortunately for the style.